frontends/concrete-python/examples/levenshtein_distance/levenshtein_distance.py

```diff
@@ -0,0 +1,218 @@
+# Computing Levenstein distance between strings, https://en.wikipedia.org/wiki/Levenshtein_distance
+
+import time
+from functools import lru_cache
+
+import numpy
+
+from concrete import fhe
+
+# Parameters to be set by the user
+max_string_length = 6
+
+
+# Module FHE
+@fhe.module()
+class MyModule:
+    @fhe.function({"x": "encrypted", "y": "encrypted"})
+    def equal(x, y):
+        return fhe.univariate(lambda x: x == 0)(x - y)
+
+    @fhe.function(
+        {
+            "is_equal": "encrypted",
+            "if_equal": "encrypted",
+            "case_1": "encrypted",
+            "case_2": "encrypted",
+            "case_3": "encrypted",
+        }
+    )
+    def mix(is_equal, if_equal, case_1, case_2, case_3):
+        min_12 = numpy.minimum(case_1, case_2)
+        min_123 = numpy.minimum(min_12, case_3)
+
+        return fhe.if_then_else(is_equal, if_equal, 1 + min_123)
+
+    # There is a single output in mix: it can go to
+    #   - input 1 of mix
+    #   - input 2 of mix
+    #   - input 3 of mix
+    #   - input 4 of mix
+    # or just be the final output
+    #
+    # There is a single output of equal, it goes to input 0 of mix
+    composition = fhe.Wired(
+        [
+            fhe.Wire(fhe.AllOutputs(equal), fhe.Input(mix, 0)),
+            fhe.Wire(fhe.AllOutputs(mix), fhe.Input(mix, 1)),
+            fhe.Wire(fhe.AllOutputs(mix), fhe.Input(mix, 2)),
+            fhe.Wire(fhe.AllOutputs(mix), fhe.Input(mix, 3)),
+            fhe.Wire(fhe.AllOutputs(mix), fhe.Input(mix, 4)),
+        ]
+    )
+
+
+# For now, we pick only small letters
+def random_letter_as_int():
+    return 97 + numpy.random.randint(26)
```

Comment:
Why do you say we are on 8b? I would say we are not, at least when we take consecutive letters, since what is important is the `equal` function, and here, it's about subtracting and checking if it's zero or not. 

But yes, if we want to move to non-consecutive letters, we can use a mapping

Reply:
When you use `==`, even with `comparison_strategy_preference=fhe.ComparisonStrategy.ONE_TLU_PROMOTED`, `equal` will have an 8b tlu (because there is no inputset evaluation for the intermediate operations) so you'll have `tlu(x - y)` and that tlu is on `8b`. We can't know if the values are consecutive in MLIR conversion. But user can encode it in their code using alphabets.

Besides, it's useful for the DNA case regardless.

Reply:
hum, but then why did you ask me to change my 

```
def equal(x, y):
        return fhe.univariate(lambda x: x == 0)(x - y)
```

which was better, with 

```
  func.func @equal(%arg0: !FHE.eint<7>, %arg1: !FHE.eint<7>) -> !FHE.eint<1> {
    %0 = "FHE.to_signed"(%arg0) : (!FHE.eint<7>) -> !FHE.esint<7>
    %1 = "FHE.to_signed"(%arg1) : (!FHE.eint<7>) -> !FHE.esint<7>
    %2 = "FHE.sub_eint"(%0, %1) : (!FHE.esint<7>, !FHE.esint<7>) -> !FHE.esint<7>
    %c2_i8 = arith.constant 2 : i8
    %3 = "FHE.mul_eint_int"(%2, %c2_i8) : (!FHE.esint<7>, i8) -> !FHE.esint<7>
    %4 = "FHE.reinterpret_precision"(%3) : (!FHE.esint<7>) -> !FHE.esint<6>
    %cst = arith.constant dense<[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]> : tensor<64xi64>
    %5 = "FHE.apply_lookup_table"(%4, %cst) : (!FHE.esint<6>, tensor<64xi64>) -> !FHE.eint<1>
    return %5 : !FHE.eint<1>
  }
```

? I get it for DNA but for the use-case I had written, it was better, and even when I move to DNA, I don't see how it would not be better

Reply:
Because `==` is only better in this case because the input format is not optimized. If it was optimized, they'd have the same performance.

Reply:
> Because == is only better in this case because the input format is not optimized. If it was optimized, they'd have the same performance.

I have not understood that sentence, is there some typos here?

Reply:
What I wanted to say is, if the input format was optimized, they would have had the same performance characteristics.

`==` is implemented in a generic way to work on all bit widths, it doesn't inspect the values in the inputset, it just checks bit width of its inputs and use that information.

But your implementation explicitly defines another intermediate that captures the continuity in a big range property in its bound measurement, so it was optimized better.

Your inputs are 7b because of the absence of alphabets, so `==` assumes they can be anything in the 7b range. But if you were using alphabets, they would be in their proper range, which is 5b, so the TLU would have been 6b, the same as the trick.

But anyway, the important thing here is that this implementation is can be massively improved using Alphabets, especially for cases like DNA, since the range is not contiguous there, `ATCG` the biggest gap here is bigger than 20, so they would be forced to be 5b, whereas they can be 2b. That's a massive difference.

Reply:
Agreed! OK, let me move to Alphabets, I wouldn't have done it but if you want yes.

Reply:
@umut-sahin I have not been able to find the python package which has Alphabet. Could you point it to me please? (I guess it's not one of those few-stars repos)

Reply:
Of course we can code that ourselves, but I understood you knew an existing standard library?

Reply:
Let me do that @umut-sahin